### PR TITLE
Philox reference implementation

### PR DESCRIPTION
**Pull request by rarutyun**
@aelizaro, @iburyl,

I needed to reopen this PR because of remote branch renaming.

### CODE REVIEW

File: rng/philox/philox.hpp [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        }
+        else {
+            int tail = z % n;
+            if (tail == 0 && state_i == 3) {
+                increment_counter(z / n);
+            }
+            else {
+                z -= num_available;
+                state_i = tail - 1;
+                increment_counter((z - 1) / n);
```
Comment (iburyl) [inline rng/philox/philox.hpp:145]: this may happen to increment by zero.

File: rng/philox/philox.hpp [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+            {
+                break;
+            }
+        }
+    }
+
+    void increment_counter(unsigned long long z)
+    {
+        for(std::size_t j = 0; j < n; j++) {
+            z += (unsigned long long)x[j];
```
Comment (iburyl) [inline rng/philox/philox.hpp:241]: z = -1; // max unsigned long long
x[0] = 1;
z += x[0]; // -> z=0; overflow, skipping zero elements!

File: rng/philox/philox.hpp [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+            for (std::size_t k = 0; k < array_size; ++k)
+            {
+                auto [mulhi, mullo] = mulhilo(v[2 * k], multipliers[k]);
+                x[2 * k + 1] = mullo;
+                x[2 * k] = mulhi ^ keys[k] ^ v[2 * k + 1];
+                keys[k] = (keys[k] + round_consts[k]) & result_mask;
+            }
+        }
+        return [&x]<std::size_t... Is>(std::index_sequence<Is...>) {
+            return std::array<result_type, n>{x[Is]...};
```
Comment (iburyl) [inline rng/philox/philox.hpp:221]: see the comments in killed review!

**Review by iburyl — CHANGES_REQUESTED** [review]

File: rng/philox/philox.hpp [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+            y = philox_generate(k, x); // see below
+            increment_counter();
+            state_i = 0;
+        }
+        return y[state_i];
+    }
+
+    void discard(unsigned long long z) {
+        std::uint32_t num_available = n - 1 - state_i;
+        if (z < num_available) {
```
[the remark this answers is not in the document]
Comment (aelizaro) [reply rng/philox/philox.hpp:134]: Fix is here: https://github.com/rarutyun/iso_cpp_papers/pull/5/files#diff-b66dd5e61a22e32b2c13d6b09bb720a5e8047a541e6d8785e9e8d25d19becfdd

[the remark this answers is not in the document]
Comment (iburyl) [reply rng/philox/philox.hpp:134]: Ok...

available = 1; state_i = 2;
z = 4;
Last brach;
tail = 0;
state_i = tail-1 ; // boom!

File: rng/philox/philox.hpp [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+    }
+
+    // equality operators
+    friend bool operator==(const philox_engine& left, const philox_engine& right) {
+        bool result = std::ranges::equal(left.x, right.x)
+            && std::ranges::equal(left.k, right.k)
+            && left.state_i == right.state_i;
+        for (auto i = left.state_i + 1; (i < n) && result; ++i)
+        {
+            result &= left.y[i] == right.y[i];
```
[the remark this answers is not in the document]
Comment (rarutyun) [reply rng/philox/philox.hpp:115]: There is no operator `&&=`. I can rewrite it as 
```cpp
result = result && (left.y[i] == right.y[i]);
```
Is it better?

File: rng/philox/philox.hpp [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+    static_assert((n & (n - 1)) == 0); // n is a power of two
+    static_assert(sizeof...(consts) == n); // check if n is even
+    static_assert(n <= 16); // any power of 2 that is <= 16
+    static_assert(w > 0 && w <= std::numeric_limits<UIntType>::digits);
+
+    static constexpr std::size_t array_size = n / 2;
+
+    using half_index_sequence = std::make_index_sequence<array_size>;
+
+    template <typename IndexSequence, template <std::size_t...> typename TransformFunction>
```
[the remark this answers is not in the document]
Comment (iburyl) [reply rng/philox/philox.hpp:32]: `internal` namespace is fine for me

File: rng/philox/philox.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (iburyl) [reply rng/philox/philox.hpp:115]: yep, it is, everything is better than doing binary ops with bool variables.

File: rng/philox/philox.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (aelizaro) [reply rng/philox/philox.hpp:134]: fixed and added a test to cover all initial positions of state_i: https://github.com/rarutyun/iso_cpp_papers/blob/07130f7165be7cb3b10d86aab8eb98ca3de9fae2/rng/philox/test.cpp#L71

File: rng/philox/philox.hpp [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+#include <ostream>
+#include <algorithm>
+
+namespace std {
+
+template <typename UIntType, std::size_t w, std::size_t n, std::size_t r, UIntType... consts>
+struct philox_engine
+{
+private:
+    static_assert(n > 0);
```
[the remark this answers is not in the document]
Comment (rarutyun) [reply rng/philox/philox.hpp]: Done

File: rng/philox/philox.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (rarutyun) [reply rng/philox/philox.hpp]: Done

File: rng/philox/philox.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (rarutyun) [reply rng/philox/philox.hpp]: Done

File: rng/philox/philox.hpp (hunk 3 above)
Comment (rarutyun) [reply rng/philox/philox.hpp:221]: Agreed to keep it.

File: rng/philox/philox.hpp (hunk 1 above)
Comment (rarutyun) [reply rng/philox/philox.hpp]: Done

File: rng/philox/philox.hpp [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+        }
+        return y[state_i];
+    }
+
+    void discard(unsigned long long z) {
+        std::uint32_t num_available = n - 1 - state_i;
+        if (z < num_available) {
+            state_i += z;
+        }
+        else {
```
[the remark this answers is not in the document]
Comment (rarutyun) [reply rng/philox/philox.hpp:138]: Rewritten (still hard to parse, to me. Would like to improve it)

File: rng/philox/philox.hpp [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+
+    void seed(result_type value = default_seed) {
+        k[0] = value & result_mask;
+        for (std::size_t j = 1; j < array_size; ++j) {
+            k[j] = 0;
+        }
+        for (std::size_t j = 0; j < n; ++j) {
+            x[j] = 0;
+        }
+        state_i = 3;
```
[the remark this answers is not in the document]
Comment (rarutyun) [reply rng/philox/philox.hpp]: Fixed.

File: rng/philox/philox.hpp (hunk 2 above)
Comment (rarutyun) [reply rng/philox/philox.hpp]: Seems like it works now. I don't like current implementation from C++ perspective (as we discussed) but we decided to apply the quickest fix to have a correct implementation and then improve it if necessary

File: rng/philox/philox.hpp [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+
+    void discard(unsigned long long z) {
+        std::uint32_t available_in_buffer = n - 1 - state_i;
+        if (z <= available_in_buffer) {
+            state_i += z;
+        }
+        else {
+            z -= available_in_buffer;
+            int tail = z % n;
+            if (tail == 0 && state_i == (n - 1)) {
```
Comment (iburyl) [inline rng/philox/philox.hpp:158]: former state_i should not influence flow at this point.
should be
```
if(tail == 0) {
  state_i = n-1;
  increment_counter(z / n);
} ...
```

File: rng/philox/philox.hpp [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+                increment_counter(z / n);
+            }
+            else {
+                state_i = tail != 0 ? tail - 1 : (n - 1);
+                if (z > 1) {
```
Comment (iburyl) [inline rng/philox/philox.hpp:163]: if (z>n)

File: rng/philox/philox.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (iburyl) [reply rng/philox/philox.hpp:155]: made two recommendations below.

File: rng/philox/philox.hpp (hunk 10 above)
Comment (rarutyun) [reply rng/philox/philox.hpp]: Done

File: rng/philox/philox.hpp (hunk 11 above)
Comment (rarutyun) [reply rng/philox/philox.hpp]: done

**Review by iburyl — APPROVED** [review]
ok

**Comment by rarutyun** [comment]
I got an approval, so it's fine to merge
